scripts/changelog.py

```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env python3
+"""
+Changelog generator
+"""
+
+import argparse
+import re
+import subprocess
+import sys
+
+import jinja2
+import requests
+from logzero import logger
+
+
+def main(args):
+    issue_keys = get_issues_from_git(args.current_release, args.new_release)
+    issues = get_issue_info(issue_keys, args.email, args.api_token)
+    if len(issues) == 0:
+        logger.warning('No issues in release')
+        return
+
+    if args.html:
+        save_html(
+            {
+                'issues': issues,
+                'tags': {
+                    'current': args.current_release,
+                    'new': args.new_release,
+                },
+            },
+            args.html,
+        )
+        if sys.platform == 'darwin':
+            subprocess.run(['open', args.html])
+    else:
+        print_changelog(args.current_release, args.new_release, issues)
+
+
+def get_issues_from_git(current_release, new_release):
+    # git log v0.9.0-rc3..v0.9.0-rc4 --oneline
+    #   | grep -E '[A-Z]+-\d+'
+    #   | perl -ne '/([A-Z]+-\d+)/; print "$1\n"'
+    #   | sort -u
+    git_log = subprocess.run(
+        ['git', 'log', f'{current_release}..{new_release}', '--oneline'],
+        capture_output=True,
+    )
+    git_log = git_log.stdout.splitlines()
+    issue_re = re.compile(r'[A-Z]+-\d+')
+    issue_keys = []
+    for msg in git_log:
+        matches = issue_re.findall(msg.decode('utf-8'))
+        for m in matches:
+            issue_keys.append(m)
+    return sorted(set(issue_keys))
+
+
+def get_issue_info(issue_keys, email, api_token):
+    issues = []
+    for key in issue_keys:
+        r = requests.get(
+            f'https://insolar.atlassian.net/rest/api/3/issue/{key}',
+            {
+                'fields': 'summary,components,status,resolution,issuetype,priority,fixVersions',
+            },
+            auth=(email, api_token)
+        )
+        try:
+            issues.append(r.json())
+        except ValueError:
+            logger.error(f'Failed to get info for issue {key}')
+    return issues
+
+
+def print_changelog(current, new, issues):
+    print(f'Changelog from {current} to {new}:')
+    for i in issues:
+        issue = i
+        print(f"{issue['key']}: {issue['fields']['summary']}")
+
+
+def prepare_context(context):
+    status_colors = {
+        'new': 'secondary',
+        'indeterminate': 'info',
+        'done': 'success',
+    }
+    for i in context['issues']:
+        key = i['fields']['status']['statusCategory']['key']
+        if key in status_colors:
+            i['fields']['status']['statusCategory']['status'] = status_colors[key]
+    return context
+
+
+def save_html(context, filename):
+    context = prepare_context(context)
+    html = render_html(context)
+    file = open(filename, 'w+')
+    file.write(html)
+    file.close()
+
+
+def render_html(context):
+    tpl = jinja2.Environment().from_string(page_tpl())
+    return tpl.render(context)
+
+
+def page_tpl():
```

Comment:
Why do you use function here? Make it constant

Reply:
✓